dissect/jffs/jffs2.py

```diff
@@ -0,0 +1,381 @@
+from __future__ import annotations
+
+import os
+import logging
+import stat
+import zlib
+from bisect import bisect_right
+from datetime import datetime
+from functools import cache, cached_property, lru_cache
+from operator import attrgetter
+from typing import BinaryIO, Iterator, Optional, Union
+
+from dissect.util import ts
+from dissect.util.compression import lzo
+from dissect.util.stream import AlignedStream, RunlistStream
+
+from dissect.jffs.c_jffs2 import DT_MAP, JFFS2_MAGIC_NUMBERS, c_jffs2
+from dissect.jffs.exceptions import (
+    Error,
+    FileNotFoundError,
+    NotADirectoryError,
+    NotASymlinkError,
+)
+
+log = logging.getLogger(__name__)
+log.setLevel(os.getenv("DISSECT_LOG_JFFS", "CRITICAL"))
+
+
+class JFFS2:
+    """Journalling Flash File System, version 2 implementation.
+
+    Currently supports the most common compression formats (``zlib``, ``lzo``, ``none`` and ``zero``).
+    CRC checksums are currently ignored. Compression formats ``rtime``, ``rubinmips``, ``copy``
+    and ``dynrubin`` are not supported.
+
+    References:
+        - https://sourceware.org/jffs2/
+        - https://github.com/torvalds/linux/blob/master/fs/jffs2/readinode.c
+    """
+
+    def __init__(self, fh: BinaryIO):
+        self.fh = fh
+
+        self._inodes: dict[int, list[tuple[c_jffs2.jffs2_raw_inode, int]]] = {}
+        self._dirents: dict[int, dict[bytes, list[DirEntry]]] = {}
+        self._lost_found: list[list[DirEntry]] = []
+
+        if (node := c_jffs2.jffs2_unknown_node(fh)).magic not in JFFS2_MAGIC_NUMBERS:
+            raise Error(f"Unknown JFFS2 magic: {node.magic:#x}")
+
+        self.inode = lru_cache(4096)(self.inode)
+
+        self.root = self.inode(inum=0x1, type=c_jffs2.DT_DIR, parent=None)
+
+        self._scan()
+        self._garbage_collect()
+
+    def inode(self, inum: int, type: Optional[int] = None, parent: Optional[INode] = None) -> INode:
+        return INode(self, inum, type, parent)
+
+    def get(self, path: Union[str, int], node: Optional[INode] = None) -> INode:
+        if isinstance(path, int):
+            return self.inode(path)
+
+        parts = path.split("/")
+        node = node or self.root
+
+        for part_num, part in enumerate(parts):
+            if not part:
+                continue
+
+            for dirent_children in self._dirents.values():
+                if dirent := dirent_children.get(part.encode()):
+                    if dirent[-1].parent_inum == node.inum:
+                        node = self.inode(dirent[-1].inum)
+                        # TODO: symlinks
+                        break
+            else:
+                raise FileNotFoundError(f"File not found: {path}")
```

Comment:
```suggestion
            if dirent := self._dirents[node.inum].get(part):
                node = self.inode(dirent[-1].inum)
                # TODO: symlinks
            else:
                raise FileNotFoundError(f"File not found: {path}")
```
I did not test this, so please verify. IIRC `self._dirents` is a mapping of `(ino) => (name of entries)`, so you should be able to index directly into the `self._dirents` dict with the current node `inum`, and then directly index into that with the current part name. If no part name exists then the path is invalid.

I did not understand why the `if dirent[-1].parent_inum == node.inum:` was necessary, but it's probably a JFFS2 specific thing I overlooked?

You can also chose to sort the directory entries by version like you currently do in `iterdir()`. In fact, it might be nice to do that during the parsing phase? Or do sorted inserts using `bisect`.

Reply:
Thanks for the suggestion, added in [3d6e94c](https://github.com/fox-it/dissect.jffs/pull/1/commits/3d6e94c7ecc4da0c5b7acf659403413ade882143). This indeed seems to work, the check on `parent_inum == node.inum` was too stringent and turned out to not be necessary.

Reply:
Thanks, that looks good. My last remaining comment would be the last sentence in the last comment:

> You can also chose to sort the directory entries by version like you currently do in iterdir(). In fact, it might be nice to do that during the parsing phase? Or do sorted inserts using bisect.

I'll leave it up to you if you wish to proceed down this road. I can implement this part if you need.

Reply:
Just so we understand correctly, you want to sort all `DirEntry`s by the `version` attribute in the list `self._dirents[pino][b"name"]`? If so, sorting would be neat, but perhaps unnecessary. Due to JFFS2 being a log/journal-based filesystem we currently assume the last `DirEntry` we encounter is also the latest version.

Reply:
I've added the sorting just in case in https://github.com/fox-it/dissect.jffs/pull/1/commits/9b4f7d315afe179097b98e34ec3c70f3465c1e28, because I see other implementations doing something similar (dropping older versions).

I don't have any huge JFFS2 images, so I'm not sure what the performance implications are. If it's large, we can do something more optimistic (only ensure the last entry is the latest, leave the rest unsorted).